app/assets/javascripts/TortoiseJS/agent/view.coffee

```diff
@@ -203,108 +204,21 @@ class TurtleDrawer
-
-      ctx.strokeStyle = netlogoColorToCSS(link.color)
-      ctx.lineWidth = if link.thickness > @view.onePixel then link.thickness else @view.onePixel
-      ctx.beginPath()
-
-      patchWidth = @view.patchWidth
-      patchHeight = @view.patchHeight
-      if wrapX and wrapY
-        # Unnecessary lines are drawn here since we're not checking to see which
-        # are actually on screen. However, browsers are better at these checks
-        # than we are and will ignore offscreen stuff. Thus, we shouldn't bother
-        # checking unless we see a consistent performance improvement. Note that
-        # at least 3 lines will be needed in the majority of cases and 4 lines
-        # are necessary in certain cases. -- BCH (3/30/2014)
-        if x1 < x2
-          if y1 < y2
-            @drawLine(x1, y1, x2 - patchWidth, y2 - patchHeight)
-            @drawLine(x1 + patchWidth, y1, x2, y2 - patchHeight)
-            @drawLine(x1 + patchWidth, y1 + patchHeight, x2, y2)
-            @drawLine(x1, y1 + patchHeight, x2 - patchWidth, y2)
-          else
-            @drawLine(x1, y1, x2 - patchWidth, y2 + patchHeight)
-            @drawLine(x1 + patchWidth, y1, x2, y2 + patchHeight)
-            @drawLine(x1 + patchWidth, y1 - patchHeight, x2, y2)
-            @drawLine(x1, y1 - patchHeight, x2 - patchWidth, y2)
-        else
-          if y1 < y2
-            @drawLine(x1, y1, x2 + patchWidth, y2 - patchHeight)
-            @drawLine(x1 - patchWidth, y1, x2, y2 - patchHeight)
-            @drawLine(x1 - patchWidth, y1 + patchHeight, x2, y2)
-            @drawLine(x1, y1 + patchHeight, x2 + patchWidth, y2)
-          else
-            @drawLine(x1, y1, x2 + patchWidth, y2 + patchHeight)
-            @drawLine(x1 - patchWidth, y1, x2, y2 + patchHeight)
-            @drawLine(x1 - patchWidth, y1 - patchHeight, x2, y2)
-            @drawLine(x1, y1 - patchHeight, x2 + patchWidth, y2)
-      else if wrapX
-        if x1 < x2
-          @drawLine(x1, y1, x2 - patchWidth, y2)
-          @drawLine(x1 + patchWidth, y1, x2, y2)
-        else
-          @drawLine(x1, y1, x2 + patchWidth, y2)
-          @drawLine(x1 - patchWidth, y1, x2, y2)
-      else if wrapY
-        if y1 < y2
-          @drawLine(x1, y1, x2, y2 - patchHeight)
-          @drawLine(x1, y1 + patchHeight, x2, y2)
-        else
-          @drawLine(x1, y1 - patchHeight, x2, y2)
-          @drawLine(x1, y1, x2, y2 + patchHeight)
-      else
-        @drawLine(x1, y1, x2, y2)
-
-    ctx.stroke()
+  drawLink: (link, turtles, wrapX, wrapY) ->
+    @linkDrawer.draw(link, turtles, wrapX, wrapY)
 
   repaint: (model) ->
     world = model.world
     turtles = model.turtles
     links = model.links
-    if world.turtleshapelist != @drawer.shapes and typeof world.turtleshapelist == "object"
-      @drawer = new CachingShapeDrawer(world.turtleshapelist)
+    if world.turtleshapelist isnt @turtleShapeDrawer.shapes and typeof world.turtleshapelist is "object"
```

Comment:
Think `typeof world.turtleshapelist is "object"` should be just `world.turtleshapelist?`. Don't know why I did it the other way before.